### ISSUE TITLE
.

### ISSUE THREAD

**Issue by YH-2023** [orrzohar/PROB#43]

**Comment by YH-2023** [comment]
![image](https://github.com/orrzohar/PROB/assets/59163676/8ab1cb06-fdd1-403a-ba45-4f861dc5cbef)
a.  I have other questions that I don't quite understand, do I have to subtract 1 from the first number I marked? When I use self.num_classes-1 in my code, the value of the loss_ce becomes very large (hundreds), and if self.num_classes does not subtract 1, the value of loss_ce is normal, I wonder if this will affect the final result? 
b.  When I enable my code to mark the number 2, the training time suddenly becomes very long when calculating the class_error (2 days becomes 5 days), why is that? 
c.  Another is that I found that the loss value during training includes class_error and cardinality_error_unscaled, as well as loss_bbox_unscaled and so on... I wonder if not adding these values to the loss, only including loss_bbox, loss_ce, loss_giou and loss_obj_ll will affect the result?
Thank you very much for your reply

**Comment by YH-2023** [comment]
![image](https://github.com/orrzohar/PROB/assets/59163676/6b75606a-0399-4769-b3b7-3e381ee42ee1)
What confuses me is that after a certain number of trainings, the value of the loss_obj does not change and stabilizes at 0.02158, what are the possible reasons for this problem? I am using a dataset from the OWOD paper

**Comment by dxjdx** [comment]
> ![image](https://user-images.githubusercontent.com/59163676/265191113-6b75606a-0399-4769-b3b7-3e381ee42ee1.png) What confuses me is that after a certain number of trainings, the value of the loss_obj does not change and stabilizes at 0.02158, what are the possible reasons for this problem? I am using a dataset from the OWOD paper

请问您用在了什么模型上呢

**Comment by dxjdx** [comment]
> ![image](https://user-images.githubusercontent.com/59163676/265191113-6b75606a-0399-4769-b3b7-3e381ee42ee1.png) What confuses me is that after a certain number of trainings, the value of the loss_obj does not change and stabilizes at 0.02158, what are the possible reasons for this problem? I am using a dataset from the OWOD paper

请问您是在什么模型上改的呢

**Comment by orrzohar** [comment]
Hi @YH-2023,

**Regarding post 1**:
If you want to add the probabilistic objectness head to a different model, you need to treat it like your other heads (bbox/cls). Specifically, yes, you need to: 
1. Define it as part of the model (in __init__), and even “clone” it (see bottom part of PROB’s init) if you are using the refinement process proposed in “deformable DETR”.
https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/models/prob_deformable_detr.py#L142
2.  You need to use it to predict the objectness per proposal. Note that the objectness head will only work with “NxD” dimensional embeddings (similar to the DETR cls head), and to use it with a different type of embedding will require some experimentation.
3. You need to incorporate it in the “forward” of the model.
https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/models/prob_deformable_detr.py#L260
4. You need to optimize it as described in my manuscript: https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/models/prob_deformable_detr.py#L417-L421

**Regarding post 2**:
a) This is because “num_classes” includes the “unknown” object logit (unless you changed it from 81 to 80), which you have no labels for, and therefore is not included in the cls head training. I have no idea if removing the “-1” will change the result — I have never tried to do this.
b) This is really legacy code, and I never used this function. My guess is that the “accuracy” function is just very slow. It is not used to compute the “total_loss” during training, so it is used for logging purposes only.
c) The losses you mentioned are not summed in training: https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/engine.py#L56  
“weight_dict” is defined as: 

https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/models/prob_deformable_detr.py#L619
  Except when you add the auxilely losses (all the _1, _2 losses):
https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/models/prob_deformable_detr.py#L625-L630


**Regarding post 3** 
This is not surprising, the objectness head is trained via alternating optimization. These are my old graphs: 
![image](https://github.com/orrzohar/PROB/assets/108689663/00810793-1186-45eb-968c-3bd287cee840)
 In general, I would be more concerned about the eval performance on OWOD (K-mAP/U-R) then the actual loss values, they are somewhat useful to supervise the training process but aren’t super critical in their absolute values. I did have more fluxuations though.

Are you using the this repository or your own custom one?

Best,  
Orr

**Comment by YH-2023** [comment]
Thank you for your prompt reply, an improvement made by a two-stage network of Detectron2 that I used during inference.
a.   I found a problem during the inference process, using a dataset from prob, but the data sampler is executed differently than the value printed out of your network.
![image](https://github.com/orrzohar/PROB/assets/59163676/295b7c56-9ba8-4c8a-97b0-8ccb2f5230b3)

![image](https://github.com/orrzohar/PROB/assets/59163676/56b8804d-f33c-4c73-8ced-09bc0401cc7b)
When I execute this code, the samples data printed out is as follows
![image](https://github.com/orrzohar/PROB/assets/59163676/0426281f-6e3d-4491-bec2-f68cd6755ece)
Here is the samples data printed out of your data:
![image](https://github.com/orrzohar/PROB/assets/59163676/a360a8b6-dd72-42b6-86a9-8edf0de5e8e0)
A little confused is why the same samples are different with the same code for the same dataset?

b.  The second serious problem is that my network does not seem to support the NestedTensor data format, how do I convert NestedTensor to the tensor format I need, is there any good way to solve this problem? Because I also tried the data format that does not apply to NestedTensor in prob, the score obtained by executing the results = postprocessors['bbox'] (outputs, orig_target_sizes) line of code is very low (0.0...).
Here's the forward part of my backbone:
![image](https://github.com/orrzohar/PROB/assets/59163676/abd95853-3bb0-4833-b58b-6f67c5f2102a)
Now the problem is that if you do not use the data sampling method in prob, the score obtained is very low, and if the data format in prob seems to be not very supportive, do you have any good suggestions?
We look forward to hearing from you.Thanks again for your help.

**Comment by orrzohar** [comment]
Hi @YH-2023 

A nested tensor is a data structure that generalizes the concept of a tensor to allow for hierarchically nested, variable-length dimensions. Unlike a regular tensor, which has fixed-size dimensions, a nested tensor can contain other tensors of different shapes as its elements. They are useful for handling data with non-uniform or hierarchical shapes.

As you can see in the image above, to get the original list of tensors, you have the ".tensor" property.

What do you mean "data sampling method in prob"?

I, unfortunately, do not really understand what you are asking. Perhaps you could rephrase?

Best,
Orr

**Comment by YH-2023** [comment]
If known classes are treated as unknown, it may also result in an increase in the unknown metrics. Are there any evaluation metrics for this scenario?

**Comment by orrzohar** [comment]
Hi @YH-2023,

Detecting known objects as "unknown" does not increase "unknown recall", and as you are limited to 100 total predictions per images will actually decrease "unknown recall" (as you have essentially "wasted" predictions) as well as K-mAP. Additionally, WI will be higher (as P_K will remain unchanged but P_K+U will go down, so WI will go up, which is bad). 

What metric of unknown object detection will be positively impacted by detecting known objects as unknown objects?

Best,
Orr

**Comment by orrzohar** [comment]
Hi @YH-2023,

It seems like you are predicting many classes to be "unknown". I would look at your model's known/unknown number of predictions. My guess is that before you made more "known" object prediction, as indicated in the first part of your question. As to why, this means that your model predicts "unknown" objects too confidently. If you are using PROB, changing the objectness temperature is one way of changing this for a trained model. The other is to change the down-weigh of the "unknown object" logit during training. 

To get the best "feel" of it, I would visualize the predictions you are making. 
Best,
Orr.

**Comment by YH-2023** [comment]
Hi @orrzohar 
Since the known is not regarded as the unknown resulting in high unknown recall, how can we balance the known and the unknown?I don't use objectness temperature in my code.
In my code, when calculating the class probability, the probability of the unknown class is much higher than the probability of the known class, is this the reason why the unknown class is overconfident?Is it possible to mitigate by reducing the weight of the newly added object branch?From the above printing your labels, there are also a lot of labels that are 80, why not there will be unknown overpredictions?
Thank you for your reply again

**Comment by orrzohar** [comment]
Hi @YH-2023 

Yes, it seems like your model is overconfident in making unknown object predictions. WIthout knowing the details of your model, it is hard to say how to mitigate it, but your suggestion of changing the weight of the "unknown object" loss is reasonable.  

The "80" = unknown objects, so you are making a lot of unknown object predcitions.

Unknown recall = number of correctly identified unknown object instances / total number of GT unknown objects

Best,
Orr

**Comment by YH-2023** [comment]
Hi @orrzohar During the reasoning process, when I execute class PostProcess, the value of the output temperature is 0.005078125. In the paper, I see it is 1.3. Why is this?

**Comment by orrzohar** [comment]
Hi @YH-2023,

I opened a new issue and answered this, please see issue https://github.com/orrzohar/PROB/issues/44

Best,
Orr

**Comment by YH-2023** [comment]
Hi @orrzohar 
What is the reason for the decrease in the number of predicted instances of known classes? Is it related to the predicted probability or the predicted prediction box? After adding the temperature, changing the value of the temperature, I found that the known classes still basically remain unchanged. At present, my known class is still at 47,At present, the number of known detected instances has decreased. so helpless.

**Comment by orrzohar** [comment]
Hi @YH-2023,

You need to visualize the detections to know that -- although I would be very surprised if it had to do with the box prediction.

When you visualize the top-100 predictions, it is good to evaluate the score of each prediction - specifically the known and unknown predictions. You will probably see that the unknown predictions are too confident.

If you want to make sure, you can "zero-out" (just multiply by 0) the unknown object logit and see the effect that has on known object prediction. It will always go up, but if you are now getting really good results you can be sure this is an issue with unknown object confidence rather than known object predictions.
 
Again, without understanding what your setup is I cannot know if there is an architectural issue or not. But you may need to play with the hyper parameters a bit. specifically:
https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/main_open_world.py#L147

and the "empty_weight" (I would just define another argparser argument and pass it through to criterion), this is just a down-weighting the unknown object/background logit during optimization, as you have a lot more unknown object predictions then known object predictions (~1:10). 
https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/models/prob_deformable_detr.py#L319

You probably will need to play with the lr_drop of the training as well:
https://github.com/orrzohar/PROB/blob/10b6518f90495e07b7baf0d1bfa353f0e583eb8e/main_open_world.py#L46

In my experiments, the known object mAP consistently increases with mAP, so perhaps your network was just not done training yet. 

Best,
Orr

**Comment by YH-2023** [comment]
Hi @orrzohar 
When I'm calculating my losses,
target_classes = torch.full(src_logits.shape[:2], self.num_classes -1,
                                     dtype=torch.int64, device=src_logits.device)
This line of code performs self.num_classes -1, then I execute the classification branch during the training process:
self.class_logits = nn.Linear(d_model, num_classes) also need to perform the operation of num_classes -1? That is, self.class_logits = nn.Linear(d_model, num_classes)--->self.class_logits = nn.Linear(d_model , num_classes-1)?

**Comment by orrzohar** [comment]
Hi @YH-2023,

No, you do not want to do num_classes-1. 

torch.full creates a tensor of shape src_logits.shape[:2], filled with self.num_classes-1, or 80. 
This discrepancy is due to pythonic indexing (which starts at 0, not 1). So, for 81 classes, the max index is 80, not 81, as you have a 0 index. 

Orr

**Comment by YH-2023** [comment]
I'll close this question first. If there are any questions, I'll open it again. Thank you again for your patient reply.
